### ISSUE TITLE
Improve support for transforms

### ISSUE THREAD

**Issue by joshdover** [elastic/package-spec#370]
We merged basic transforms support in https://github.com/elastic/package-spec/pull/307, but there are several outstanding and related problems that remain to be solved:
- [ ] Better upgrade strategy for transforms
	* Rebuilding transform destination indices can be very expensive, but is necessary in some cases
	* Should avoid it in obvious upgrade scenarios (eg. a visualization was updated)
	* Updating the frequency or delay is supported by the transform API, Fleet could leverage this to avoid re-building transforms unnecessarily so.
	* Will definitely need to destroy and re-create a transform when the aggregation or fields have changed. This could be improved, but likely lower priority at this time.
- [ ] Better privilege model for transforms
    * `kibana_system` is currently used to install all transforms. This is necessary in the Endpoint case where the package must be upgraded along with Kibana. CSP may need this too. 
	* See https://github.com/elastic/package-spec/issues/293
- [ ] Space-specific transforms
- [ ] Enforce the data stream naming convention for transform src and dest indices
- [ ] https://github.com/elastic/package-spec/issues/202

**Comment by sophiec20** [comment]
Thanks for a good summary of improvements.

Re space-specific transforms .. Future changes to the transforms privileges model have a strong dependency on a better understanding of the plans for a unified platform security model. 

Within the current limitations of the kibana and elasticsearch split privileges model - transforms supports Spaces as follows. The transform destination index is a normal elasticsearch index. When using Kibana Spaces, create a Kibana Data View on the destination index for the required Space. Spaces are supported. The long running transform persistent task is not space aware (in the same way that an elasticsearch index or an ILM policy is not space aware). This is by design. 

Ideally this item should be de-scoped (or shown as long term) from Fleet integrations work as this has caused confusion in the past.

**Comment by szeitlin** [comment]
@sophiec20 Thanks for clarifying. I agree, I don't think space-specific transforms is a high priority for us, though installing package assets with the appropriate privileges I'm told has been/could be a blocker. @joshdover wrote some detailed suggestions re: API keys here, but I definitely don't know enough about the trade-offs re: auto-upgrade or how important that should be. https://github.com/elastic/package-spec/issues/293#issuecomment-1192736955

**Comment by sophiec20** [comment]
@szeitlin Your comment mostly refers to `Better privilege model for transforms` (which should remain in this list).

**Comment by szeitlin** [comment]
> @szeitlin Your comment mostly refers to `Better privilege model for transforms` (which should remain in this list).

@sophiec20 Correct, but it doesn't seem to be a ticket or linked to something that's obviously roadmapped to be working on in 8.4 or 8.5? Is there a plan for how to address that issue? It looked like it was still under discussion, and from what I've read so far, it seems potentially complex to decide how we want the privilege model to work, nevermind how to implement it.

**Comment by joshdover** [comment]
> doesn't seem to be a ticket or linked to something that's obviously roadmapped to be working on in 8.4 or 8.5? Is there a plan for how to address that issue?

I've opened a new issue for this: https://github.com/elastic/kibana/issues/137278

**Comment by joshdover** [comment]
Fleet now supports installing transforms based on the new spec, thanks to @qn895's work in https://github.com/elastic/kibana/issues/134321. Next steps as I understand it:
- (maybe) Revisit the upgrade behavior, we likely need to at least support the naive approach of deleting the destination index and rebuilding the transform. See discussion here: https://github.com/elastic/kibana/issues/134321#issuecomment-1252892113
- https://github.com/elastic/kibana/issues/137278
	- First, align on technical definition, including how this impacts other asset types, like ILM. We need a common, scalable approach.
	- Implement, again no one is yet assigned.
